### PR TITLE
Python 3 compatibility

### PR DESCRIPTION
**Pull request by ajratner**
Primarily follows steps from [python-future.org](http://python-future.org/quickstart.html). Currently waiting on a change in Numbskull to pass generative model tests (https://github.com/HazyResearch/numbskull/pull/51). Will assign reviewers when passing all tests in both Python2 + 3

### CODE REVIEW

**Comment by ajratner** [comment]
TODO:

- [ ] Go through and correct inefficient `iteritems` and `list(range` bits from conversion script
- [ ] Why is `SparseLogisticRegression` seeming to fail tests??

**Comment by ajratner** [comment]
Note: Gold labels for Intro tutorial seem to be misaligned in Python 3... everything runs but the scores are all messed up... presumably some sort of string parsing issue

**Comment by ajratner** [comment]
Notes on errors in Intro 1 parsing:
- One issue was sentence didn't get split due to `\x` chars, which made the count of people in it too high... in general think just some parsing issue...

**Comment by ajratner** [comment]
Passing Python 3 travis tests- stuck on a dumb install thing for Python 2 checks...

**Comment by ajratner** [comment]
@henryre (and @dhimmel?): No obligation, but if either of you want to do a quick pass, much appreciated!  Otherwise I'll merge in Thursday am.  Thanks!

**Comment by henryre** [comment]
Looking now

File: snorkel/udf.py [hunk 1]
```diff
@@ -1,8 +1,16 @@
...
+from future import standard_library
+standard_library.install_aliases()
+from builtins import *
+
 from multiprocessing import Process, JoinableQueue
 try:
     from queue import Empty
 except:
-    from Queue import Empty
+    from queue import Empty
```
Comment (henryre) [inline snorkel/udf.py]: :/

File: tree_structs.py [hunk 2]
```diff
@@ -103,7 +103,7 @@ def scrub(s):
 def corenlp_to_xmltree_sub(s, dep_parents, rid=0):
   i = rid - 1
   attrib = {}
-  N = len(dep_parents)
+  N = len(list(dep_parents))
```
Comment (henryre) [inline tree_structs.py]: ?

File: tree_structs.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def scrub(s):
 def corenlp_to_xmltree_sub(s, dep_parents, rid=0):
```
Comment (henryre) [inline tree_structs.py]: Also why isn't this file somewhere else?

**Comment by ajratner** [comment]
@henryre Travis is testing on Python 2 and 3!  I'll look over the other comments in a bit, thanks!

**Comment by henryre** [comment]
@ajratner I think it's testing on Python 2 or Python 3 but may be wrong. Do you have a log that shows everything being run twice?

**Comment by ajratner** [comment]
@henryre Yeah there are two "jobs" per Travis build- you can see if you go into Travis.  The "matrix" stuff is just if you want to consider many different combinations of configurations

**Comment by henryre** [comment]
@ajratner noice

File: snorkel/contrib/brat/brat.py [hunk 4]
```diff
@@ -236,7 +242,7 @@ def map_annotations(self, session, annotation_dir, candidates, symmetric_relatio
 
         n, N = len(mapped_cands), len(missed) + len(mapped_cands)
         p = len(mapped_cands)/ float(N)
-        print>>sys.stderr,"Mapped {}/{} ({:2.0f}%) of BRAT labels to candidates".format(n,N,p*100)
+        print("{} Mapped {}/{} ({:2.0f}%) of BRAT labels to candidates".format(sys.stderr,n,N,p*100))
```
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/contrib/brat/brat.py]: yeah... this was extreme laziness on my part.  `contrib` not guaranteed... but I'll fix

File: snorkel/learning/utils.py [hunk 5]
```diff
@@ -298,55 +300,6 @@ def print_scores(ntp, nfp, ntn, nfn, title='Scores'):
 
 
 
-############################################################
```
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/learning/utils.py]: Yes but was causing Python 3 issues, and we'd been meaning to delete anyway

File: snorkel/learning/utils.py [hunk 6]
```diff
@@ -463,6 +418,9 @@ def _fit_st(self, X_valid, Y_valid, b=0.5, beta=1,
             run_stats.append(list(param_vals) + list(run_scores))
             if run_score > run_score_opt or k == 0:
                 model.save(model_name=model_name, save_dir=self.save_dir)
+                # Also save a separate file for easier access
+                model.save(model_name="{0}_best".format(model.name),
```
[the remark this answers is not in the document]
Comment (ajratner) [reply snorkel/learning/utils.py]: The parsing is slightly different between python versions --> the results are slightly different --> the best model is not necessarily the same.  This was the cleanest way to ensure stability for tests, as well as a good (out of scope) feature

File: tree_structs.py (hunk 3 above)
Comment (ajratner) [reply tree_structs.py]: 🤷‍♂️

File: tree_structs.py (hunk 2 above)
Comment (ajratner) [reply tree_structs.py]: 🤷‍♂️

**Comment by ajratner** [comment]
@henryre made all requested comments save those I responded to- lmk if further needed

File: tutorials/intro/util.py [hunk 7]
```diff
@@ -45,4 +51,4 @@ def load_external_labels(session, candidate_class, annotator_name='gold'):
 
     # Reload annotator labels
     reload_annotator_labels(session, candidate_class, annotator_name, split=1, filter_label_split=False)
-    reload_annotator_labels(session, candidate_class, annotator_name, split=2, filter_label_split=False)
+    reload_annotator_labels(session, candidate_class, annotator_name, split=2, filter_label_split=False)
```
Comment (dhimmel) [inline tutorials/intro/util.py]: Add newline back

File: snorkel/annotations.py [hunk 8]
```diff
@@ -367,10 +375,12 @@ def load_matrix(matrix_class, annotation_key_class, annotation_class, session,
...
     # it takes 8min to go throuh 245M rows (pretty fast).
+    for res in session.execute(annot_select_query):
+        # NOTE: The order of return seems to be switched in Python 3???
+        # Either way, make sure the order is set here explicitly!
+        cid, kid, val = res.candidate_id, res.key_id, res.value
 
-    for val, cid, kid in session.execute(annot_select_query):
-
-        if cid in cid_to_row and kid in kid_to_col:
+        if (cid in cid_to_row) and (kid in kid_to_col):
```
Comment (dhimmel) [inline snorkel/annotations.py]: The parentheses in this if statement are not needed as `in` is higher [precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence) than `and`.

File: snorkel/contrib/brat/utils.py [hunk 9]
```diff
@@ -14,6 +19,6 @@ def download(url, outfname):
         with open(outfname, "wb") as f:
             f.write(data.read())
     except HTTPError, e:
-        print "HTTP Error:", e.code, url
+        print("HTTP Error: {} {}".format(e.code, url))
     except URLError, e:
-        print "URL Error:", e.reason, url
\ No newline at end of file
+        print("URL Error: {} {}".format(e.reason, url))
```
Comment (dhimmel) [inline snorkel/contrib/brat/utils.py]: Add newline at end of file

File: snorkel/learning/structure/__init__.py [hunk 10]
```diff
@@ -1,6 +1,8 @@
 """
 Subpackage for learning the structures of models.
 """
+from __future__ import absolute_import
+
 from .gen_learning import DependencySelector
 from .synthetic import generate_model, generate_label_matrix
-from .utils import *
+from .utils import *
```
Comment (dhimmel) [inline snorkel/learning/structure/__init__.py]: Newline

File: snorkel/learning/utils.py [hunk 11]
```diff
@@ -261,10 +260,10 @@ def summary_score(self, test_marginals, **kwargs):
         """
         error_sets = self.score(test_marginals, display=False, **kwargs)
         if len(error_sets) == 4:
-            _, _, f1 = binary_scores_from_counts(*map(len, error_sets))
+            _, _, f1 = binary_scores_from_counts(*list(map(len, error_sets)))
```
Comment (dhimmel) [inline snorkel/learning/utils.py]: No need to convert to list. Can unpack a generator:

```python
print(*range(6))
```

File: snorkel/learning/utils.py [hunk 12]
```diff
@@ -261,10 +260,10 @@ def summary_score(self, test_marginals, **kwargs):
...
             return f1, "F1 Score"
         else:
-            nc, ninc = map(len, error_sets)
+            nc, ninc = list(map(len, error_sets))
```
Comment (dhimmel) [inline snorkel/learning/utils.py]: List not needed. The following works:

```
a, b = range(2)
```

File: snorkel/learning/utils.py [hunk 13]
```diff
@@ -373,7 +323,7 @@ def __init__(self, model_class, parameter_dict, X_train, Y_train=None,
         model_class_params={}, model_hyperparams={}, save_dir='checkpoints'):
         self.model_class        = model_class
         self.parameter_dict     = parameter_dict
-        self.param_names        = parameter_dict.keys()
+        self.param_names        = list(parameter_dict.keys())
```
Comment (dhimmel) [inline snorkel/learning/utils.py]: equivalent to `list(parameter_dict)`

File: snorkel/learning/utils.py [hunk 14]
```diff
@@ -425,9 +375,11 @@ def _fit_st(self, X_valid, Y_valid, b=0.5, beta=1,
             for pn, pv in zip(self.param_names, param_vals):
                 hps[pn] = pv
             print("=" * 60)
-            NUMTYPES = [float, int, np.float64]
+            NUMTYPES = (float, int, np.float64)
```
Comment (dhimmel) [inline snorkel/learning/utils.py]: delete parentheses

File: snorkel/lf_helpers.py [hunk 15]
```diff
@@ -135,7 +141,7 @@ def contains_token(c, tok, attrib='words', case_sensitive=False):
     except:
         spans = [c]
     f = (lambda w: w) if case_sensitive else (lambda w: w.lower())
-    return f(tok) in set(chain.from_iterable(map(f, span.get_attrib_tokens(attrib))
+    return f(tok) in set(chain.from_iterable(list(map(f, span.get_attrib_tokens(attrib)))
```
Comment (dhimmel) [inline snorkel/lf_helpers.py]: remove list

File: snorkel/matchers.py [hunk 16]
```diff
@@ -30,7 +36,7 @@ def _check_opts(self):
         Checks for unsupported opts, throws error if found
         NOTE: Must be called _after_ init()
         """
-        for opt in self.opts.keys():
+        for opt in list(self.opts.keys()):
```
Comment (dhimmel) [inline snorkel/matchers.py]: Remove list

File: snorkel/parser/__init__.py [hunk 17]
```diff
@@ -1,6 +1,8 @@
+from __future__ import absolute_import
+
 from .corenlp import *
 from .corpus_parser import *
 from .doc_preprocessors import *
 from .parser import *
 from .spacy_parser import *
-from .rule_parser import *
+from .rule_parser import *
```
Comment (dhimmel) [inline snorkel/parser/__init__.py]: newline

File: snorkel/parser/parser.py [hunk 18]
```diff
@@ -18,15 +24,15 @@ def to_unicode(self, text):
         :param text:
         :return:
         '''
-        text = text.encode('utf-8', 'error')
-        # Python 2
+        # Python 2 only -- should fail on LookupError in Python 3
```
Comment (dhimmel) [inline snorkel/parser/parser.py]: Perhaps make the logic more explicit:

```
if sys.version_info[0] < 3:
```

Don't allow the python 2 steps to be run on python 3?

**Comment by ajratner** [comment]
@dhimmel thanks so much for the great comments!!  Made all requested changes. Happy holidays / NYE!!

**Comment by ajratner** [comment]
@dhimmel @henryre will plan to merge *tomorrow AM* if tests pass and no further comments- thanks for the help!!

**Pull request merged**